app/app.go

```diff
@@ -1096,14 +1114,92 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
 	return app.mm.InitGenesis(ctx, app.appCodec, genesisState, app.genesisImportConfig)
 }
 
-func (app *App) PrepareProposalHandler(_ sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+func (app *App) PrepareProposalHandler(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	// Get all pending bundles for this height
+	fmt.Println("Preparing proposal for height", ctx.BlockHeight())
+
+	bundleRes, err := app.MevKeeper.PendingBundles(sdk.WrapSDKContext(ctx), &mevtypes.QueryPendingBundlesRequest{})
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("Found", len(bundleRes.Bundles), "pending bundles from mevkeeper")
+
+	maxTxBytes := req.MaxTxBytes
+	var selectedTxs [][]byte
+	bundleTxMap := make(map[string]struct{})
+
+	// First, add any system transactions (governance, etc)
+	for _, tx := range req.Txs {
+		if app.isSystemTx(tx) {
+			selectedTxs = append(selectedTxs, tx)
+		}
+	}
+
+	// Next, add bundle transactions with highest priority
+	for _, bundle := range bundleRes.Bundles {
+		// Skip bundles not meant for this height
+		if bundle.BlockNum != uint64(ctx.BlockHeight()) {
+			continue
+		}
+
+		// Calculate total size of this bundle
+		bundleSize := int64(0)
```

Comment:
nit: I'd do `var bundleSize int64` rather than this. I'd also not worry about `int` being smaller than `int64` but that last one is aesthetics.